src/authorization/ParentContainerReader.ts

```diff
@@ -0,0 +1,100 @@
+import type { PermissionMap, Permission, PermissionSet, AccessMap } from './permissions/Permissions';
+import { AccessMode } from './permissions/Permissions';
+
+/**
+ * Determines `delete` and `create` permissions for those resources that need it
+ * by making sure the parent container has the required permissions.
+ *
+ * Create requires `append` permissions on the parent container.
+ * Delete requires `write` permissions on both the parent container and the resource itself.
+ */
+export class ParentContainerReader extends PermissionReader {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly reader: PermissionReader;
+  private readonly identifierStrategy: IdentifierStrategy;
+
+  public constructor(reader: PermissionReader, identifierStrategy: IdentifierStrategy) {
+    super();
+    this.reader = reader;
+    this.identifierStrategy = identifierStrategy;
+  }
+
+  public async handle({ requestedModes, credentials }: PermissionReaderInput): Promise<PermissionMap> {
+    const containerMap = this.findParents(requestedModes);
+
+    const updatedMap = modify(new IdentifierSetMultiMap(requestedModes), { add: containerMap.values() });
+    const result = await this.reader.handleSafe({ requestedModes: updatedMap, credentials });
+
+    for (const [ identifier, [ container ]] of containerMap) {
+      this.logger.debug(`Determining ${identifier.path} create and delete permissions based on ${container.path}`);
+      result.set(identifier, this.updatePermission(result.get(identifier), result.get(container)));
+    }
+    return result;
+  }
+
+  /**
+   * Finds the identifiers for which we need parent permissions.
+   * Values are the parent identifier and the permissions they need.
+   */
+  private findParents(requestedModes: AccessMap): IdentifierMap<[ResourceIdentifier, Set<AccessMode>]> {
+    const containerMap = new IdentifierMap<[ResourceIdentifier, Set<AccessMode>]>();
+    for (const [ identifier, modes ] of requestedModes.entrySets()) {
+      if (modes.has(AccessMode.create) || modes.has(AccessMode.delete)) {
+        const container = this.identifierStrategy.getParentContainer(identifier);
+        containerMap.set(identifier, [ container, this.getParentModes(modes) ]);
+      }
+    }
+    return containerMap;
+  }
+
+  /**
+   * Determines which permissions are required on the parent container.
+   */
+  private getParentModes(modes: ReadonlySet<AccessMode>): Set<AccessMode> {
+    const containerModes: Set<AccessMode> = new Set();
+    if (modes.has(AccessMode.create)) {
+      containerModes.add(AccessMode.append);
+    }
+    if (modes.has(AccessMode.delete)) {
+      containerModes.add(AccessMode.write);
+    }
+    return containerModes;
+  }
+
+  /**
+   * Determines the create and delete permissions for the given permission set based on those if its parent container.
+   */
+  private updatePermission(permissionSet?: PermissionSet, containerSet?: PermissionSet): PermissionSet {
```

Comment:
```suggestion
  private withContainerPermission(permissionSet?: PermissionSet, containerSet?: PermissionSet): PermissionSet {
```
or something (update what? update how?)